File: snapkit-fortran/src/spectral.f90
!> Spectral analysis — entropy, Hurst exponent, autocorrelation.
!> Fortran 2008 port of snapkit spectral module.
!> Column-major friendly, pure where possible, BLAS-ready layout.
module snapkit_spectral
    implicit none
    private
    public :: entropy, autocorrelation, hurst_exponent, spectral_summary
    public :: spectral_batch, spectral_result

    real(8), parameter :: LOG2_INV = 1.4426950408889634d0  ! 1/ln(2)
    real(8), parameter :: INV_E    = 0.36787944117144233d0 ! 1/e

    type :: spectral_result
        real(8) :: entropy_bits    = 0.0d0
        real(8) :: hurst           = 0.5d0
        real(8) :: autocorr_lag1   = 0.0d0
        real(8) :: autocorr_decay  = 0.0d0
        logical :: is_stationary   = .false.
    end type

contains

    !> Shannon entropy via histogram binning (bits).
    pure function entropy(data, bins) result(h)
        real(8), intent(in) :: data(:)
        integer, intent(in), optional :: bins
        real(8) :: h

        integer :: n, nb, i, idx, c
        real(8) :: min_val, max_val, inv_range, p
        integer, allocatable :: counts(:)

        n = size(data)
        if (n < 2) then
            h = 0.0d0
            return
        end if

        nb = 10
        if (present(bins)) nb = bins

        ! Find min/max
        min_val = data(1)
        max_val = data(1)
        do i = 2, n
            if (data(i) < min_val) min_val = data(i)
            if (data(i) > max_val) max_val = data(i)
        end do

        if (max_val == min_val) then
            h = 0.0d0
            return
        end if

        inv_range = dble(nb) / (max_val - min_val)
        allocate(counts(nb))
        counts = 0

        do i = 1, n
            idx = int((data(i) - min_val) * inv_range) + 1
            if (idx > nb) idx = nb
            if (idx < 1) idx = 1
            counts(idx) = counts(idx) + 1
        end do

        h = 0.0d0
        do i = 1, nb
            c = counts(i)
            if (c > 0) then
                p = dble(c) / dble(n)
                h = h - p * log(p) * LOG2_INV
            end if
        end do
    end function

    !> Normalized autocorrelation up to max_lag.
    function autocorrelation(data, max_lag_in) result(acf)
        real(8), intent(in) :: data(:)
        integer, intent(in), optional :: max_lag_in
        real(8), allocatable :: acf(:)

        integer :: n, max_lag, lag, t, limit
        real(8) :: mean_val, inv_n, r0, inv_r0, rk
        real(8), allocatable :: centered(:)

        n = size(data)
        if (n < 2) then
            allocate(acf(1))
            acf(1) = 1.0d0
            return
        end if

        max_lag = n / 2
        if (present(max_lag_in)) max_lag = min(max_lag_in, n - 1)

        inv_n = 1.0d0 / dble(n)

        ! Center the data
        mean_val = sum(data) * inv_n
        allocate(centered(n))
        centered = data - mean_val

        ! Variance
        r0 = dot_product(centered, centered) * inv_n
        if (r0 == 0.0d0) then
            allocate(acf(max_lag + 1))
            acf(1) = 1.0d0
            acf(2:) = 0.0d0
            return
        end if

        inv_r0 = 1.0d0 / r0
        allocate(acf(max_lag + 1))

        do lag = 0, max_lag
            rk = 0.0d0
            limit = n - lag
            do t = 1, limit
                rk = rk + centered(t) * centered(t + lag)
            end do
            acf(lag + 1) = rk * inv_n * inv_r0
        end do
    end function

    !> Hurst exponent via R/S analysis.
    function hurst_exponent(data) result(h)
        real(8), intent(in) :: data(:)
        real(8) :: h

        integer :: n, i, size_val, num_sub, start, j, t
        integer :: n_pts, rs_count
        real(8) :: inv_n, mean_val, inv_size, sub_mean
        real(8) :: running, cum_min, cum_max, r_val, var_val, d_val
        real(8) :: rs_sum, avg_rs
        real(8) :: sum_x, sum_y, sum_xy, sum_x2, denom
        real(8), allocatable :: centered(:), sizes(:), rs_values(:)
        integer, allocatable :: test_sizes(:)
        integer :: nsizes, prev_s

        n = size(data)
        if (n < 20) then
            h = 0.5d0
            return
        end if

        inv_n = 1.0d0 / dble(n)
        mean_val = sum(data) * inv_n
        allocate(centered(n))
        centered = data - mean_val

        ! Build geometric test sizes: 16, 32, 48, 64, ...
        nsizes = 0
        size_val = 16
        prev_s = 0
        do while (size_val <= n / 2)
            nsizes = nsizes + 1
            prev_s = size_val
            if (size_val * 2 <= n / 2) then
                size_val = size_val * 2
            else
                size_val = int(dble(size_val) * 1.5d0)
            end if
            if (size_val == prev_s) exit
        end do

        if (nsizes == 0) then
            if (n >= 8) nsizes = 1
        end if
        if (nsizes == 0) then
            h = 0.5d0
            return
        end if

        allocate(test_sizes(nsizes))
        nsizes = 0
        size_val = 16
        prev_s = 0
        do while (size_val <= n / 2)
            nsizes = nsizes + 1
            test_sizes(nsizes) = size_val
            prev_s = size_val
            if (size_val * 2 <= n / 2) then
                size_val = size_val * 2
            else
                size_val = int(dble(size_val) * 1.5d0)
            end if
            if (size_val == prev_s) exit
        end do

        ! Compute R/S for each size
        allocate(sizes(nsizes), rs_values(nsizes))
        n_pts = 0

        do i = 1, nsizes
            size_val = test_sizes(i)
            if (size_val < 4 .or. size_val > n) cycle

            num_sub = n / size_val
            if (num_sub < 1) cycle

            inv_size = 1.0d0 / dble(size_val)
            rs_sum = 0.0d0
            rs_count = 0

            do j = 1, num_sub
                start = (j - 1) * size_val + 1
                sub_mean = sum(centered(start:start+size_val-1)) * inv_size

                ! Cumulative deviations with inline min/max
                running = 0.0d0
                cum_min = 0.0d0
                cum_max = 0.0d0
                var_val = 0.0d0
                do t = 1, size_val
                    d_val = centered(start + t - 1) - sub_mean
                    running = running + d_val
                    if (running < cum_min) cum_min = running
                    if (running > cum_max) cum_max = running
                    var_val = var_val + d_val * d_val
                end do

                r_val = cum_max - cum_min
                var_val = var_val * inv_size

                if (var_val > 1.0d-20) then
                    rs_sum = rs_sum + r_val / sqrt(var_val)
                    rs_count = rs_count + 1
                end if
            end do

            if (rs_count > 0) then
                avg_rs = rs_sum / dble(rs_count)
                if (avg_rs > 0.0d0) then
                    n_pts = n_pts + 1
                    sizes(n_pts) = dble(size_val)
                    rs_values(n_pts) = avg_rs
                end if
            end if
        end do

        if (n_pts < 2) then
            h = 0.5d0
            return
        end if

        ! Linear regression on log-log
        sum_x = 0.0d0; sum_y = 0.0d0; sum_xy = 0.0d0; sum_x2 = 0.0d0
        do i = 1, n_pts
            d_val = log(sizes(i))
            r_val = log(rs_values(i))
            sum_x = sum_x + d_val
            sum_y = sum_y + r_val
            sum_xy = sum_xy + d_val * r_val
            sum_x2 = sum_x2 + d_val * d_val
        end do

        denom = dble(n_pts) * sum_x2 - sum_x * sum_x
        if (denom == 0.0d0) then
            h = 0.5d0
            return
        end if

        h = (dble(n_pts) * sum_xy - sum_x * sum_y) / denom

        ! Clamp to [0, 1]
        if (h < 0.0d0) h = 0.0d0
        if (h > 1.0d0) h = 1.0d0
    end function

    !> Full spectral summary of a signal.
    function spectral_summary(data, bins, max_lag_in) result(sr)
        real(8), intent(in) :: data(:)
        integer, intent(in), optional :: bins, max_lag_in
        type(spectral_result) :: sr

        real(8), allocatable :: acf(:)
        integer :: i, nb, ml

        nb = 10
        if (present(bins)) nb = bins

        sr%entropy_bits = entropy(data, nb)
        sr%hurst = hurst_exponent(data)

        if (present(max_lag_in)) then
            acf = autocorrelation(data, max_lag_in)
        else
            acf = autocorrelation(data)
        end if

        sr%autocorr_lag1 = 0.0d0
        if (size(acf) > 1) sr%autocorr_lag1 = acf(2)

        ! Find decay lag (first lag where |acf| < 1/e)
        sr%autocorr_decay = dble(size(acf))
        do i = 2, size(acf)
            if (abs(acf(i)) < INV_E) then
                sr%autocorr_decay = dble(i - 1)  ! 0-indexed lag
                exit
            end if
        end do

        ! Stationarity check
        sr%is_stationary = (sr%hurst >= 0.4d0 .and. sr%hurst <= 0.6d0) .and. &
                           (abs(sr%autocorr_lag1) < 0.3d0)
    end function

    !> Batch spectral summary for multiple time series.
    !> Each column of data_2d is one time series (column-major layout).
    subroutine spectral_batch(data_2d, results, bins, max_lag_in)
        real(8), intent(in)  :: data_2d(:,:)
        type(spectral_result), intent(out) :: results(:)
        integer, intent(in), optional :: bins, max_lag_in
        integer :: i

        do i = 1, size(data_2d, 2)
            if (present(bins) .and. present(max_lag_in)) then
                results(i) = spectral_summary(data_2d(:,i), bins, max_lag_in)
            else if (present(bins)) then
                results(i) = spectral_summary(data_2d(:,i), bins)
            else if (present(max_lag_in)) then
                results(i) = spectral_summary(data_2d(:,i), max_lag_in=max_lag_in)
            else
                results(i) = spectral_summary(data_2d(:,i))
            end if
        end do
    end subroutine

end module snapkit_spectral

File: snapkit-zig/src/spectral.zig
//! Spectral analysis — entropy, Hurst R/S exponent, autocorrelation.
//!
//! No hidden allocations. All scratch buffers provided by the caller.

const std = @import("std");
const types = @import("types.zig");
const SpectralSummary = types.SpectralSummary;

/// Shannon entropy via histogram binning.
/// `counts` must be a slice of length `bins`.
pub fn entropy(data: []const f64, bins: usize, counts: []usize) f64 {
    const n = data.len;
    if (n < 2 or bins < 1) return 0.0;

    // Find min/max
    var min_val = data[0];
    var max_val = data[0];
    for (data[1..]) |x| {
        if (x < min_val) min_val = x;
        if (x > max_val) max_val = x;
    }
    if (max_val == min_val) return 0.0;

    const inv_range = @as(f64, @floatFromInt(bins)) / (max_val - min_val);

    // Zero counts
    @memset(counts[0..bins], 0);

    for (data) |x| {
        var idx: usize = @intFromFloat((x - min_val) * inv_range);
        if (idx >= bins) idx = bins - 1;
        counts[idx] += 1;
    }

    const inv_n = 1.0 / @as(f64, @floatFromInt(n));
    const inv_log2 = 1.0 / @log(2.0);
    var h: f64 = 0.0;
    for (counts[0..bins]) |c| {
        if (c > 0) {
            const p = @as(f64, @floatFromInt(c)) * inv_n;
            h -= p * @log(p) * inv_log2;
        }
    }
    return h;
}

/// Normalized autocorrelation.
/// `out` must have length at least max_lag + 1.
/// Returns the number of lags written.
pub fn autocorrelation(data: []const f64, max_lag: usize, out: []f64) usize {
    const n = data.len;
    if (n < 2) {
        if (out.len > 0) out[0] = 1.0;
        return 1;
    }

    const effective_lag = @min(max_lag, n - 1);
    if (out.len < effective_lag + 1) return 0;

    const inv_n = 1.0 / @as(f64, @floatFromInt(n));

    // Compute mean
    var mean: f64 = 0.0;
    for (data) |x| mean += x;
    mean *= inv_n;

    // Compute variance (r0)
    var r0: f64 = 0.0;
    for (data) |x| {
        const d = x - mean;
        r0 += d * d;
    }
    r0 *= inv_n;

    if (r0 == 0.0) {
        out[0] = 1.0;
        for (out[1 .. effective_lag + 1]) |*o| o.* = 0.0;
        return effective_lag + 1;
    }

    const inv_r0 = 1.0 / r0;

    for (0..effective_lag + 1) |lag| {
        var rk: f64 = 0.0;
        const limit = n - lag;
        for (0..limit) |t| {
            const d1 = data[t] - mean;
            const d2 = data[t + lag] - mean;
            rk += d1 * d2;
        }
        out[lag] = rk * inv_n * inv_r0;
    }
    return effective_lag + 1;
}

/// Hurst exponent via R/S analysis.
pub fn hurstExponent(data: []const f64) f64 {
    const n = data.len;
    if (n < 20) return 0.5;

    const inv_n = 1.0 / @as(f64, @floatFromInt(n));

    // Compute mean
    var mean: f64 = 0.0;
    for (data) |x| mean += x;
    mean *= inv_n;

    // We need centered data — use an allocator-free approach:
    // Process subseries directly from the original data, subtracting sub_mean.
    // This avoids allocating a centered array.

    // Build geometric progression of test sizes
    var test_sizes: [32]usize = undefined;
    var n_sizes: usize = 0;
    var s: usize = 16;
    while (s <= n / 2 and n_sizes < 32) {
        test_sizes[n_sizes] = s;
        n_sizes += 1;
        const next = s * 2;
        s = if (next <= n / 2) next else @intFromFloat(@as(f64, @floatFromInt(s)) * 1.5);
        if (s <= test_sizes[n_sizes - 1]) break;
    }
    if (n_sizes == 0 and n >= 8) {
        test_sizes[0] = n / 4;
        n_sizes = 1;
    }

    var log_sizes: [32]f64 = undefined;
    var log_rs: [32]f64 = undefined;
    var n_pts: usize = 0;

    for (test_sizes[0..n_sizes]) |size| {
        if (size < 4 or size > n) continue;

        const num_sub = n / size;
        if (num_sub < 1) continue;

        const inv_size = 1.0 / @as(f64, @floatFromInt(size));
        var rs_sum: f64 = 0.0;
        var rs_count: usize = 0;

        for (0..num_sub) |i| {
            const start = i * size;
            const sub = data[start .. start + size];

            // Compute sub-mean
            var sub_mean: f64 = 0.0;
            for (sub) |x| sub_mean += x;
            sub_mean *= inv_size;

            // Cumulative deviations with min/max tracking
            var running: f64 = 0.0;
            var cum_min: f64 = 0.0;
            var cum_max: f64 = 0.0;
            for (sub) |x| {
                running += x - sub_mean;
                if (running < cum_min) cum_min = running;
                if (running > cum_max) cum_max = running;
            }
            const R = cum_max - cum_min;

            // Variance
            var var_val: f64 = 0.0;
            for (sub) |x| {
                const d = x - sub_mean;
                var_val += d * d;
            }
            var_val *= inv_size;

            if (var_val > 1e-20) {
                rs_sum += R / @sqrt(var_val);
                rs_count += 1;
            }
        }

        if (rs_count > 0) {
            const avg_rs = rs_sum / @as(f64, @floatFromInt(rs_count));
            if (avg_rs > 0.0) {
                log_sizes[n_pts] = @log(@as(f64, @floatFromInt(size)));
                log_rs[n_pts] = @log(avg_rs);
                n_pts += 1;
            }
        }
    }

    if (n_pts < 2) return 0.5;

    // Linear regression on log-log
    var sx: f64 = 0.0;
    var sy: f64 = 0.0;
    var sxy: f64 = 0.0;
    var sx2: f64 = 0.0;
    for (0..n_pts) |i| {
        const lx = log_sizes[i];
        const ly = log_rs[i];
        sx += lx;
        sy += ly;
        sxy += lx * ly;
        sx2 += lx * lx;
    }

    const n_pts_f = @as(f64, @floatFromInt(n_pts));
    const denom = n_pts_f * sx2 - sx * sx;
    if (denom == 0.0) return 0.5;

    var h = (n_pts_f * sxy - sx * sy) / denom;
    if (h < 0.0) h = 0.0;
    if (h > 1.0) h = 1.0;
    return h;
}

/// Full spectral summary of a signal.
/// `acf_buf` must be length at least max_lag + 1.
/// `counts_buf` must be length at least bins.
pub fn spectralSummary(
    data: []const f64,
    bins: usize,
    max_lag: usize,
    acf_buf: []f64,
    counts_buf: []usize,
) SpectralSummary {
    const h = entropy(data, bins, counts_buf);
    const hurst_val = hurstExponent(data);

    const effective_lag = if (max_lag > 0) max_lag else data.len / 2;
    const acf_len = autocorrelation(data, effective_lag, acf_buf);

    const acf_lag1: f64 = if (acf_len > 1) acf_buf[1] else 0.0;

    // Find decay lag (where |acf| < 1/e)
    const threshold: f64 = 0.36787944117144233; // 1/e
    var decay_lag: f64 = @as(f64, @floatFromInt(acf_len));
    for (1..acf_len) |i| {
        if (@abs(acf_buf[i]) < threshold) {
            decay_lag = @as(f64, @floatFromInt(i));
            break;
        }
    }

    const is_stationary = (hurst_val >= 0.4 and hurst_val <= 0.6) and @abs(acf_lag1) < 0.3;

    return .{
        .entropy_bits = h,
        .hurst = hurst_val,
        .autocorr_lag1 = acf_lag1,
        .autocorr_decay = decay_lag,
        .is_stationary = is_stationary,
    };
}

// ── C-compatible exports ──

export fn sk_entropy(data: [*]const f64, n: usize, bins: usize) f64 {
    // Use a static buffer for small bin counts
    var counts_buf: [256]usize = undefined;
    const effective_bins = @min(bins, 256);
    return entropy(data[0..n], effective_bins, &counts_buf);
}

export fn sk_hurst_exponent(data: [*]const f64, n: usize) f64 {
    return hurstExponent(data[0..n]);
}

// ── Tests ──

test "entropy — constant signal" {
    const data = [_]f64{ 1.0, 1.0, 1.0, 1.0 };
    var counts: [10]usize = undefined;
    const h = entropy(&data, 10, &counts);
    try std.testing.expectApproxEqAbs(@as(f64, 0.0), h, 1e-10);
}

test "entropy — uniform signal" {
    // With 10 bins and 10 uniformly spaced points, entropy ≈ log2(10)
    const data = [_]f64{ 0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0 };
    var counts: [10]usize = undefined;
    const h = entropy(&data, 10, &counts);
    try std.testing.expect(h > 3.0); // log2(10) ≈ 3.32
}

test "autocorrelation — lag 0 is 1.0" {
    const data = [_]f64{ 1.0, 2.0, 3.0, 4.0, 5.0 };
    var acf: [5]f64 = undefined;
    const len = autocorrelation(&data, 4, &acf);
    try std.testing.expectEqual(@as(usize, 5), len);
    try std.testing.expectApproxEqAbs(@as(f64, 1.0), acf[0], 1e-10);
}

test "autocorrelation — constant signal" {
    const data = [_]f64{ 5.0, 5.0, 5.0, 5.0 };
    var acf: [4]f64 = undefined;
    const len = autocorrelation(&data, 3, &acf);
    try std.testing.expectEqual(@as(usize, 4), len);
    try std.testing.expectApproxEqAbs(@as(f64, 1.0), acf[0], 1e-10);
    // All other lags should be 0 for constant signal
    for (acf[1..len]) |val| {
        try std.testing.expectApproxEqAbs(@as(f64, 0.0), val, 1e-10);
    }
}

test "Hurst exponent — returns 0.5 for small data" {
    const data = [_]f64{ 1.0, 2.0, 3.0 };
    const h = hurstExponent(&data);
    try std.testing.expectEqual(@as(f64, 0.5), h);
}

test "Hurst exponent — reasonable range for random-ish data" {
    // Generate a simple sequence
    var data: [100]f64 = undefined;
    for (&data, 0..) |*d, i| {
        d.* = @sin(@as(f64, @floatFromInt(i)) * 0.1);
    }
    const h = hurstExponent(&data);
    try std.testing.expect(h >= 0.0 and h <= 1.0);
}

test "spectral summary — basic" {
    var data: [100]f64 = undefined;
    for (&data, 0..) |*d, i| {
        d.* = @sin(@as(f64, @floatFromInt(i)) * 0.3);
    }
    var acf_buf: [50]f64 = undefined;
    var counts: [10]usize = undefined;
    const summary = spectralSummary(&data, 10, 25, &acf_buf, &counts);
    try std.testing.expect(summary.entropy_bits >= 0.0);
    try std.testing.expect(summary.hurst >= 0.0 and summary.hurst <= 1.0);
    try std.testing.expect(summary.autocorr_decay >= 0.0);
}
